### DISCUSSION TITLE
Passing a TypeScript Generic to the store's type

### DISCUSSION THREAD

**Discussion by zorzysty** [pmndrs/zustand#362] [Q&A] [answered]
I need to pass a TS generic to the type describing the store, but I can't figure out how to pass it through the `create` function.

I'm hoping to achieve something similar to the code below, but here of course the `Bear` inside `create<State2<Bear>>` is not defined:
```ts
// inside the store file
type State<Bear> = {
  bears: Bear[]
  increase: (newBear: Bear) => void
}
​
export const useStore = create<State<Bear>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] })),
}))
​
// and in a file using the store
type Bear = {
  kind: "Grizzly" | "Polar"
  age: number
}
​
const bears = useStore<Bear>((state) => state.bears)
```
I tried to wrap the result of `create` in a function that accepts a generic, which made TS happy, but in the runtime I of course I got "Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component.(...)". This can be mitigated by invoking state by `useState()(state => state.bears)`, but it doesn't seem right.

Inb4, I can't declare the `Bear` type in the store file as in reality I'm getting it as a generic as well and I simply must pass it along to the store.

Is there a way to do it properly?

**Comment by dai-shi** [comment, marked as answer]
I don't know if there's a TS-only way. If we can use a wrapper, something like this?

```ts
const useStoreUnknown = create<State<unknown>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] })),
}))

export const useStore = <Bear>(selector StateSelector<State<Bear>>): State<Bear> => useStoreUnknown(selector)
```

Or, maybe this works?

```ts
export const useStore = create<State<unknown>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] })),
}))

  // in component
  const bears = useStore<State<Bear>>((state) => state.bears)
```

**Reply by zorzysty** [reply]
Thank you for the answer @dai-shi! I think the first solution is quite promising, but I couldn't quite get it to work.

This was my best attempt:
```ts
type State<T> = {
  bears: T[]
  increase: (newBear: T) => void
}

const useStoreUnknown = create<State<unknown>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] })),
}))

export const useStore = <T extends Record<string, unknown>>(
  selector: StateSelector<State<T>, State<T>>
): State<T> => useStoreUnknown(selector) // here `selector` has an error:
/*
TS2345: Argument of type 'StateSelector<State<T>, State<T>>' is not assignable to parameter of type 'StateSelector<State<unknown>, State<T>>'. 
  Type 'State<unknown>' is not assignable to type 'State<T>'. 
    Types of property 'bears' are incompatible.     
     Type 'unknown[]' is not assignable to type 'T[]'.      
      Type 'unknown' is not assignable to type 'T'.         
       'T' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.
*/
```
And in a component:
```ts
const bears = useStore<Bear>((state) => state.bears) // here state.bears has an error:
/* TS2739: Type 'Bear[]' is missing the following properties from type 'State<Bear>': bears, increase */
```

I guess the problem is with the second generic passed to the `StateSelector` or the return type of `useStore`. For both I tried `State<T>` and `UseStore<State<T>>`, but nothing worked. 

Do you have any idea what am I doing wrong?

**Reply by dai-shi** [reply]
I didn't try it on my end and am not very certain, but it may require explicit type casting or the use of `any` instead of `unknown`.

**Reply by zorzysty** [reply]
Thank you very much @dai-shi! It finally worked with this approach:

```ts
type State<T> = {
  bears: T[]
  increase: (newBear: T) => void
}

const useStoreAny = create<State<any>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] })),
}))

export const useStore = <T extends Record<string | number | symbol, unknown>>(
  selector: StateSelector<State<T>, any>
): State<T> => useStoreAny(selector)
```

and in component:
```ts
const bears = useStore<Bear>((state) => state.bears)
```

**Reply by zorzysty** [reply]
I just noticed, there's still a problem 🤔
Although the type of `state.bears` in the selector is inferred correctly, the type of `bears` const is `State<T>`. And when I remove the explicit return type, it gets passed through from the second argument of `StateSelector` which is `any` in this case.
I think the second argument of `StateSelector` needs to be something more specific, but I cannot figure it out.

BTW, I got rid of one `any` by inline-ing useStoreAny:
```ts
type State<T> = {
  bears: T[]
  increase: (newBear: T) => void
}

export const useStore = <T extends Record<string | number | symbol, unknown>>(
  selector: StateSelector<State<T>, any>
) =>
  create<State<T>>((set) => ({
    bears: [],
    increase: (newBear) =>
      set((state) => ({ bears: [...state.bears, newBear] })),
  }))(selector)
```

**Reply by dai-shi** [reply]
Inlining is not good. store (useStore) should be created outside React.

How about this?
```ts
export const useStore = <T extends Record<string | number | symbol, unknown>, U>(
  selector: StateSelector<State<T>, U>
): State<T> => useStoreAny(selector)
```

**Reply by zorzysty** [reply]
With this approach you must explicitly declare the return value you expect when calling `useStore` - which technically works, but means loosing type-safety (you're effectively overriding the type of the return value).
I created a CodeSandbox: https://codesandbox.io/s/serene-bas-0b3f8?file=/src/App.tsx where you can quickly check it yourself.

And thanks for the tip about moving store creation outside of React :)

**Reply by dai-shi** [reply]
Ah ha, yeah, it requires to specify the return value type...

I don't think it's possible without double arrows, which is not good for hooks.

Maybe this one is better.
https://codesandbox.io/s/pedantic-kare-p09ec?file=/src/store.ts

```tsx
export type UseStore<T> = UseStoreOrig<State<T>>;

export const useStoreUnknown = create<State<unknown>>((set) => ({
  bears: [],
  increase: (newBear) => set((state) => ({ bears: [...state.bears, newBear] }))
}));

const bearsD = (useStoreUnknown as UseStore<Bear>)((state) => state.bears);
```

**Reply by zorzysty** [reply]
Yeah, this will work 😎
Thank you @dai-shi! You are awesome!

BTW, do you think it's technically possible to make it a built-in feature of zustand - without the need for wrappers, unknowns and explicit type casting? If so, I can create a feature request and maybe even give PR a shot if somebody points me in the right direction.

**Reply by dai-shi** [reply]
Hm, I'd say the last solution is more like a built-in feature.

If you or somebody has any idea to improve typing, suggestions are welcome.
